### PR TITLE
[CodeCamp2023-503] Add the DDQ algorithm to mmdetection

### PR DESCRIPTION
**Pull request by jiongjiongli**
Task: [CodeCamp2023-503](https://github.com/open-mmlab/OpenMMLabCamp/discussions/503)

## Motivation

Refer to the open source code [DDQ](https://github.com/jshilong/DDQ) and add support for training this model in mmdetection.

## Modification

1. Added DDQ DETR config files: 
    - configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py
    - configs/ddq_detr/ddq-detr-5scale_r50_8xb2-12e_coco.py
    - configs/ddq_detr/ddq-detr-4scale_swinl_8xb2-30e_coco.py
3. Added classes:
    - DDQDETR
    - DDQDETRHead
    - DDQTransformerDecoder
    - DDQAuxLoss
    - TopkHungarianAssigner

### Installation

```
module load python/3.8.12-gcc-4.8.5-jbm
module load cuda/11.1.0-gcc-4.8.5-67q
module load gcc/9.3.0-gcc-4.8.5-bxl

mkdir ~/ddq/new
cd ~/ddq/new/

git clone -b jiongjiongli/ddq_detr https://github.com/jiongjiongli/mmdetection.git mmdetection

cd ~/ddq/new/mmdetection/
python -m venv  ~/ddq/new/detr_py38
source ~/ddq/new/detr_py38/bin/activate

pip install --upgrade pip
pip install cython numpy \
matplotlib pycocotools scipy shapely six terminaltables \
cityscapesscripts imagecorruptions scikit-learn \
pytest-runner \
ninja psutil \
addict numpy packaging Pillow pyyaml yapf coverage lmdb onnx onnxoptimizer onnxruntime pytest PyTurboJPEG scipy tifffile \
termcolor rich \
dadaptation lion-pytorch parameterized \
ipython

pip install torch==1.9.0+cu111 torchvision==0.10.0+cu111 -f https://download.pytorch.org/whl/torch_stable.html
pip install mmengine==0.7.1
pip install mmcv==2.0.0rc4 -f https://download.openmmlab.com/mmcv/dist/cu111/torch1.9/index.html

cd ~/ddq/new/mmdetection/
pip install -v -e .

```

### Train

```

module load python/3.8.12-gcc-4.8.5-jbm
module load cuda/11.1.0-gcc-4.8.5-67q
module load gcc/9.3.0-gcc-4.8.5-bxl

source ~/ddq/new/detr_py38/bin/activate

cd ~/ddq/new/mmdetection/

PORT=50132 sh tools/train.sh configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py 8 ./exp/ddq_detr

```

### Test

```
module load python/3.8.12-gcc-4.8.5-jbm
module load cuda/11.1.0-gcc-4.8.5-67q
module load gcc/9.3.0-gcc-4.8.5-bxl

source ~/ddq/new/detr_py38/bin/activate

cd ~/ddq/new/mmdetection/

PORT=50136 sh tools/dist_test.sh  configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py $pretrain_model_path 8

```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
4. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMPreTrain.
5. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by PhoenixZ810** [comment]
in ~ddq/new/mmdetection/mmdet/models/detectors/ddq_detr.py, Row158:
`_, keep_idxs = batched_nms(single_proposals, single_scores,
                                       torch.ones(len(single_scores)),
                                       self.cache_dict['dqs_cfg'])`
there will be error: 
**RuntimeError:indices should be either on cpu or on the same device as the indexed tensor (cpu)** 
should change to：
`_, keep_idxs = batched_nms(single_proposals, single_scores,
                                       torch.ones(len(single_scores)).cuda(),
                                       self.cache_dict['dqs_cfg'])`

File: mmdet/models/layers/transformer/ddq_detr_layers.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            _, keep_idxs = batched_nms(single_proposals[ori_index],
+                                       single_scores[ori_index],
+                                       torch.ones(len(ori_index)),
+                                       self.cache_dict['dqs_cfg'])
+
+            real_keep_index = ori_index[keep_idxs]
+
+            attn_mask = torch.ones_like(dis_mask[0]).bool()
+            # such a attn_mask give best result
+            attn_mask[real_keep_index] = False
```
Comment (hhaAndroid) [inline mmdet/models/layers/transformer/ddq_detr_layers.py:74]: 这个地方要非常重要，要说清楚这个是十字架模式，而并不是某个 dist query 和 某几个 dist query 不交互，而是挑选出来的 dist query 和所有 query 进行交互，其余 query 不进行交互

File: mmdet/models/losses/ddq_detr_aux_loss.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                                         cls_scores.size(-1),
+                                         dtype=torch.long)
+            label_weights = bbox_preds.new_zeros(
+                num_valid_anchors, dtype=torch.float)
+            norm_alignment_metrics = bbox_preds.new_zeros(
+                num_valid_anchors, dtype=torch.float)
+            return (labels, label_weights, bbox_targets,
+                    norm_alignment_metrics)
+
+        assign_result = self.assigner.assign(cls_scores, bbox_preds, gt_bboxes,
```
Comment (hhaAndroid) [inline mmdet/models/losses/ddq_detr_aux_loss.py:245]: 1 和 6这两个参数不要写死吧，最好写到初始化里面，不然很奇怪

File: mmdet/models/task_modules/assigners/topk_hungarian_assigner.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        if num_gt > 0:
+            # assign 0 by default
+            assigned_gt_inds = pred_scores.new_full((num_bboxes, ),
+                                                    0,
+                                                    dtype=torch.long)
+            select_cost = all_cost
+
+            topk = min(self.topk, int(len(select_cost) / num_gt))
+
+            repeat_select_cost = select_cost[...,
```
Comment (hhaAndroid) [inline mmdet/models/task_modules/assigners/topk_hungarian_assigner.py:127]: 注明下这句话的实际意义：假设 query 是 900,gt 是3，之前 cost 是 900x3，这样一对一匹配时候就只是三个正样本，现在假设copy 4 遍，就变成 900x12了，这样就变成了 一个 gt 对应 4 个预测

File: mmdet/models/utils/misc.py [hunk 4]
```diff
@@ -650,3 +650,47 @@ def unfold_wo_center(x, kernel_size: int, dilation: int) -> Tensor:
...
+    if input_tensor.dim() > 1:
+        padding = input_tensor.new_zeros(
+            num_padding, *input_tensor.size()[1:], dtype=input_tensor.dtype)
+    else:
+        padding = input_tensor.new_zeros(num_padding, dtype=input_tensor.dtype)
+    output_tensor = torch.cat([input_tensor, padding], dim=0)
+    return output_tensor
+
+
+def align_tensor(inputs, max_len=None):
```
Comment (hhaAndroid) [inline mmdet/models/utils/misc.py:678]: 加一下 typehint

File: mmdet/models/utils/misc.py [hunk 5]
```diff
@@ -650,3 +650,47 @@ def unfold_wo_center(x, kernel_size: int, dilation: int) -> Tensor:
         dim=2)
 
     return unfolded_x
+
+
+def padding_to(input_tensor, max_len=300):
```
Comment (hhaAndroid) [inline mmdet/models/utils/misc.py:655]: 加一下 typehint

File: configs/ddq_detr/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# DDQ
+
+[Dense Distinct Query for End-to-End Object Detection](https://arxiv.org/abs/2303.12776)
+
+## Abstract
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply configs/ddq_detr/README.md:5]: 好的，已经插入。

File: configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50')),
+    neck=dict(
+        type='ChannelMapper',
+        in_channels=[512, 1024, 2048],
+        kernel_size=1,
+        out_channels=256,
+        act_cfg=None,
+        norm_cfg=dict(type='GN', num_groups=32),
+        num_outs=4),
+    encoder=dict(
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py:33]: 已添加

File: configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        num_layers=6,
+        layer_cfg=dict(
+            self_attn_cfg=dict(embed_dims=256, num_levels=4,
+                               dropout=0.0),  # 0.1 for DeformDETR
+            ffn_cfg=dict(
+                embed_dims=256,
+                feedforward_channels=2048,  # 1024 for DeformDETR
+                ffn_drop=0.0))),  # 0.1 for DeformDETR
+    decoder=dict(
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py:42]: 已添加

File: configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            type='HungarianAssigner',
+            match_costs=[
+                dict(type='FocalLossCost', weight=2.0),
+                dict(type='BBoxL1Cost', weight=5.0, box_format='xywh'),
+                dict(type='IoUCost', iou_mode='giou', weight=2.0)
+            ])),
+    test_cfg=dict(max_per_img=300))
+
+train_pipeline = [
+    dict(type='LoadImageFromFile', backend_args={{_base_.backend_args}}),
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py:89]: 已修改

File: mmdet/models/layers/transformer/ddq_detr_layers.py (hunk 1 above)
Comment (jiongjiongli) [reply mmdet/models/layers/transformer/ddq_detr_layers.py:74]: 已添加

File: mmdet/models/losses/ddq_detr_aux_loss.py (hunk 2 above)
Comment (jiongjiongli) [reply mmdet/models/losses/ddq_detr_aux_loss.py:245]: 已删除。

File: mmdet/models/task_modules/assigners/topk_hungarian_assigner.py (hunk 3 above)
Comment (jiongjiongli) [reply mmdet/models/task_modules/assigners/topk_hungarian_assigner.py:127]: 已添加

File: mmdet/models/utils/misc.py (hunk 4 above)
Comment (jiongjiongli) [reply mmdet/models/utils/misc.py:678]: 已添加

File: mmdet/models/utils/misc.py (hunk 5 above)
Comment (jiongjiongli) [reply mmdet/models/utils/misc.py:655]: 已添加

File: configs/ddq_detr/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+## Results and Models
+
+|      Model      | Backbone | Lr schd | Augmentation | box AP(val) |                      Config                      |   Download   |
+| :-------------: | :------: | :-----: | :----------: | :---------: | :----------------------------------------------: | :----------: |
+| DDQ DETR-4scale |   R-50   |   12e   |     DETR     |    51.3     | [config](./ddq-detr-4scale_r50_8xb2-12e_coco.py) | model \| log |
+| DDQ DETR-5scale |   R-50   |   12e   |     DETR     |    52.1     |                      config                      | model \| log |
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply configs/ddq_detr/README.md]: 可以直接用

**Comment by jiongjiongli** [comment]
> in ~ddq/new/mmdetection/mmdet/models/detectors/ddq_detr.py, Row158: `_, keep_idxs = batched_nms(single_proposals, single_scores, torch.ones(len(single_scores)), self.cache_dict['dqs_cfg'])` there will be error: **RuntimeError:indices should be either on cpu or on the same device as the indexed tensor (cpu)** should change to： `_, keep_idxs = batched_nms(single_proposals, single_scores, torch.ones(len(single_scores).device=single_scores.get_device()), self.cache_dict['dqs_cfg'])`

Fixed this issue, thanks!

File: tests/test_models/test_dense_heads/test_ddq_detr_head.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+from mmdet.models.dense_heads import DDQDETRHead
+from mmdet.structures import DetDataSample
+
+
+class TestDDQDETRHead(TestCase):
+
+    def test_ddq_detr_head_loss(self):
+        """Tests DDQDETRHead loss when truth is empty and non-empty."""
+        num_classes = 80
```
Comment (hhaAndroid) [inline tests/test_models/test_dense_heads/test_ddq_detr_head.py:17]: 这些参数都不需要这么大，只是单元测试而已

Comment (hhaAndroid) [reply tests/test_models/test_dense_heads/test_ddq_detr_head.py:17]: Num_class 改成2，num_dn_queries=10，类似缩放就行

File: tests/test_models/test_layers/test_transformer.py [hunk 12]
```diff
@@ -502,3 +503,21 @@ def test_detr_transformer_encoder_decoder():
                     act_cfg=dict(type='ReLU', inplace=True)))))
     assert len(DetrTransformerEncoder(**config).layers) == 6
     assert DetrTransformerEncoder(**config)
+
+
+def test_ddq_transformer_decoder():
+    config = ConfigDict(
+        num_layers=6,
```
Comment (hhaAndroid) [inline tests/test_models/test_layers/test_transformer.py:510]: 改成2

File: mmdet/models/task_modules/assigners/topk_hungarian_assigner.py (hunk 3 above)
Comment (hhaAndroid) [reply mmdet/models/task_modules/assigners/topk_hungarian_assigner.py:133]: 注释加到这行代码前面

File: mmdet/models/detectors/ddq_detr.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        proposals = enc_outputs_coord_unact.sigmoid()
+        proposals = bbox_cxcywh_to_xyxy(proposals)
+        scores = enc_outputs_class.max(-1)[0].sigmoid()
+
+        if self.training:
+            dense_proposals = dense_enc_outputs_coord_unact.sigmoid()
+            dense_proposals = bbox_cxcywh_to_xyxy(dense_proposals)
+            dense_scores = dense_enc_outputs_class.max(-1)[0].sigmoid()
+
+        num_imgs = len(scores)
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:145]: 已添加

File: mmdet/models/detectors/ddq_detr.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+        if self.training:
+            # -1 is the aux head for the encoder
+            dense_enc_outputs_class = self.bbox_head.cls_branches[-1](
+                output_memory)
+            dense_enc_outputs_coord_unact = self.bbox_head.reg_branches[-1](
+                output_memory) + output_proposals
+
+        topk = self.num_queries
+        dense_topk = int(topk * 1.5)
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:134]: 已添加

File: mmdet/models/detectors/ddq_detr.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+        query = []
+
+        dense_topk_score = []
+        dense_topk_coords_unact = []
+        dense_query = []
+
+        for img_id in range(num_imgs):
+            single_proposals = proposals[img_id]
+            single_scores = scores[img_id]
+            _, keep_idxs = batched_nms(single_proposals, single_scores,
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py]: 已添加

File: mmdet/models/detectors/ddq_detr.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+
+        if self.training:
+            dense_proposals = dense_enc_outputs_coord_unact.sigmoid()
+            dense_proposals = bbox_cxcywh_to_xyxy(dense_proposals)
+            dense_scores = dense_enc_outputs_class.max(-1)[0].sigmoid()
+
+        num_imgs = len(scores)
+        topk_score = []
+        topk_coords_unact = []
+        query = []
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:148]: 已添加

File: mmdet/models/detectors/ddq_detr.py [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+                                        [dense_keep_idxs][:dense_topk])
+                dense_topk_coords_unact.append(
+                    dense_enc_outputs_coord_unact[img_id][dense_keep_idxs]
+                    [:dense_topk])
+
+            topk_score.append(enc_outputs_class[img_id][keep_idxs][:topk])
+            topk_coords_unact.append(
+                enc_outputs_coord_unact[img_id][keep_idxs][:topk])
+
+            map_memory = self.query_map(memory[img_id].detach())
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:182]: 已添加

File: mmdet/models/detectors/ddq_detr.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+        topk_coords_unact = topk_coords_unact.detach()
+
+        if self.training:
+            dn_label_query, dn_bbox_query, dn_mask, dn_meta = \
+                self.dn_query_generator(batch_data_samples)
+            query = torch.cat([dn_label_query, query], dim=1)
+            reference_points = torch.cat([dn_bbox_query, topk_coords_unact],
+                                         dim=1)
+            ori_size = dn_mask.size(-1)
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:214]: 已添加

File: mmdet/models/dense_heads/ddq_detr_head.py [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+        End-to-End Object Detection.
+
+    Code is modified from the `official github repo
+        <https://github.com/jshilong/DDQ>`_.
+
+    More details can be found in the `paper
+        <https://arxiv.org/abs/2303.12776>`_ .
+
+    Args:
+        aux_num_pos (int): Number of positive targets assigned to a
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/dense_heads/ddq_detr_head.py:31]: 已修改

File: configs/ddq_detr/README.md [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+![ddq_arch](https://github.com/open-mmlab/mmdetection/assets/33146359/5ca9f11b-b6f3-454f-a2d1-3009ee337bbc)
+
+## Results and Models
+
+|      Model      | Backbone | Lr schd | Augmentation | box AP(val) |                       Config                       |   Download   |
+| :-------------: | :------: | :-----: | :----------: | :---------: | :------------------------------------------------: | :----------: |
+| DDQ DETR-4scale |   R-50   |   12e   |     DETR     |    51.3     |  [config](./ddq-detr-4scale_r50_8xb2-12e_coco.py)  | model \| log |
+| DDQ DETR-5scale |   R-50   |   12e   |     DETR     |    52.1     |  [config](./ddq-detr-5scale_r50_8xb2-12e_coco.py)  | model \| log |
```
Comment (hhaAndroid) [inline configs/ddq_detr/README.md:16]: 权重和 log 我已经上传了地址是：

https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq_detr_5scale_coco_1x.pth
https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq_detr_5scale_coco_1x_20230319_103307.log

https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq_detr_swinl_30e.pth
https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq_detr_swinl_30e_20230316_221721_20230318_143554.log

你填到对应位置去。并且注明：第二个和第三个是自己来自官方ddq，mmdet 并没有重新训练

File: mmdet/models/dense_heads/ddq_detr_head.py [hunk 21]
```diff
@@ -0,0 +1,540 @@
...
+            train_cfg=dict(
+                assigner=dict(type='TopkHungarianAssigner', topk=aux_num_pos),
+                alpha=1,
+                beta=6), )
+
+    def _init_layers(self) -> None:
+        """Initialize classification branch and regression branch of aux head
+        for dense queries."""
+        super(DDQDETRHead, self)._init_layers()
+        # aux head for dense queries on encoder feature map
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/dense_heads/ddq_detr_head.py:48]: 已添加

File: mmdet/models/dense_heads/ddq_detr_head.py [hunk 22]
```diff
@@ -0,0 +1,540 @@
...
+        self.cls_branches.append(copy.deepcopy(self.cls_branches[-1]))
+        self.reg_branches.append(copy.deepcopy(self.reg_branches[-1]))
+
+        # self.num_pred_layer is 7
+        # aux head for dense queries in decoder
+        self.aux_cls_branches = nn.ModuleList([
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/dense_heads/ddq_detr_head.py:54]: 已添加

File: configs/ddq_detr/README.md (hunk 20 above)
Comment (hhaAndroid) [reply configs/ddq_detr/README.md:16]: 第一个权重已经上传
https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq-detr-4scale_r50_8xb2-12e_coco/ddq-detr-4scale_r50_8xb2-12e_coco_20230809_170711-42528127.pth
https://download.openmmlab.com/mmdetection/v3.0/ddq/ddq-detr-4scale_r50_8xb2-12e_coco/ddq-detr-4scale_r50_8xb2-12e_coco_20230809_170711.log.json

Comment (hhaAndroid) [reply configs/ddq_detr/README.md:16]: mAP 是 51.4

File: configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
+_base_ = [
```
Comment (hhaAndroid) [inline configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py:1]: configs/ddq_detr 这个文件夹重命名为 configs/ddq

File: mmdet/models/dense_heads/ddq_detr_head.py [hunk 24]
```diff
@@ -0,0 +1,540 @@
...
+            Defaults to 4.
+    """
+
+    def __init__(self, *args, aux_num_pos=4, **kwargs):
+        super(DDQDETRHead, self).__init__(*args, **kwargs)
+        self.aux_loss_for_dense = DDQAuxLoss(
+            train_cfg=dict(
+                assigner=dict(type='TopkHungarianAssigner', topk=aux_num_pos),
+                alpha=1,
+                beta=6), )
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/dense_heads/ddq_detr_head.py:42]: 已修改

File: mmdet/models/detectors/ddq_detr.py [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+            single_proposals = proposals[img_id]
+            single_scores = scores[img_id]
+            _, keep_idxs = batched_nms(single_proposals, single_scores,
+                                       torch.ones(len(single_scores)),
+                                       self.cache_dict['dqs_cfg'])
+            if self.training:
+                dense_single_proposals = dense_proposals[img_id]
+                dense_single_scores = dense_scores[img_id]
+                # sort according the score
+                _, dense_keep_idxs = batched_nms(
```
[the remark this answers is not in the document]
Comment (jiongjiongli) [reply mmdet/models/detectors/ddq_detr.py:166]: 已添加

File: tests/test_models/test_dense_heads/test_ddq_detr_head.py (hunk 11 above)
Comment (jiongjiongli) [reply tests/test_models/test_dense_heads/test_ddq_detr_head.py:17]: 已修改

File: tests/test_models/test_layers/test_transformer.py (hunk 12 above)
Comment (jiongjiongli) [reply tests/test_models/test_layers/test_transformer.py:510]: 已修改

File: configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py (hunk 23 above)
Comment (jiongjiongli) [reply configs/ddq_detr/ddq-detr-4scale_r50_8xb2-12e_coco.py:1]: 已修改

File: configs/ddq_detr/README.md (hunk 20 above)
Comment (jiongjiongli) [reply configs/ddq_detr/README.md:16]: 已添加

File: mmdet/models/task_modules/assigners/topk_hungarian_assigner.py (hunk 3 above)
Comment (jiongjiongli) [reply mmdet/models/task_modules/assigners/topk_hungarian_assigner.py:133]: 加到了repeat前面可以吗： 
            topk = min(self.topk, int(len(select_cost) / num_gt))

            # Repeat the ground truth `topk` times to perform 1-to-k gt-pred
            #   matching. For example, if `num_pred` = 900, `num_gt` = 3, then
            #   there are only 3 gt-pred pairs in sum for 1-1 matching.
            #   However, for 1-k gt-pred matching, if `topk` = 4, then each
            #   gt is assigned 4 unique predictions, so there would be 12
            #   gt-pred pairs in sum.
            repeat_select_cost = select_cost[...,
                                             None].repeat(1, 1, topk).view(
                                                 select_cost.size(0), -1)

**Review by hhaAndroid — APPROVED** [review]

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @jiongjiongli,

We'd like to express our appreciation for your valuable contributions to the mmdetection. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

**Pull request merged**